rfcs/007 - Peer review infra changes.md

```diff
@@ -0,0 +1,125 @@
+- Allow data-only, `up` migrations
+- Allow plain SQL or have an introspective ORM
+- Free and ideally open source
+- Baseline/ignore certain migrations
+- Be resilient to `iasql_ugprade`
+
+Data migrations require being able to baseline migrations to avoid duplicate record violations on `iasql_upgrade`. Flyway comes closest to meeting all our requirements. However, it is not possible to baseline/ignore existing migrations on their free product.
+
+We could configure `iasql_upgrade` to preserve a `migrations` table and make the name of the table configurable, but the baselining issue would remain as `iasql_upgrade` can change the schema which renders existing migrations on past IaSQL versions invalid. As a result, we need an opinionated solution to unblock peer review of infra changes.
+
+## Proposal
+
+Take inspiration from [RFC 006](./006%20-%20Replicate%20changes%20between%20staging%20and%20prod%20RFC.md) and create a how-to guide in the documentation much like the guide on [how to replicate changes across infra environments](https://iasql.com/docs/envs/) along with an example repository that walks users through how to code review their changes using **convention over configuration** such that the workflow:
+- works with any cloud identity and access management structure
+- works with `iasql_upgrade`
+- works with any source control solution without the need for special access to it
+- preserves the ability to `iasql_preview` changes in development and get the type-safety feedback from the dashboard editor
+
+The workflow simply suggests users run `iasql_begin`, perform all the necessary changes, and before doing `iasql_commit` they manually put the output of `iasql_preview` and the SQL generated thus far into a markdown file within a repository of their liking. Once the PR is approved by their peers and landed, they can manually run `iasql_commit`. We can require a parameter for `iasql_commit(message)` which is a string attached as metadata for the transaction that can be set to be the PR URL. Best practices are thus enforced using convention and not configuration with a complete reference of what was done via the IaSQL audit log if necessary.
+
+The required work in the IaSQL engine is associating entries in the audit log with specific transactions and adding a `get_sql_for_transaction()` that returns the SQL for the current transaction so users don't have to provide a timestamp for `get_sql_since()`. Additionally, we can modify transactions so they are active past 30 minutes and do not expire at a given point in time as users are making changes to the database.
```

Comment:
I added it per @aguillenv's request. I see your point. I will remove it from this RFC since to me this does not seem like a requirement for it and seems even somewhat orthogonal.

Reply:
There is something to be said about looking at this more closely, though, and how the user-defined transactions and the cron's internal transactions ought to have differing behaviors.

1. If a cron gets "stuck" (which could be due to a crash or a restart) we *don't* want to roll back the state it was operating on, while if a manual `iasql_commit` blows up we probably do.
2. If a transaction is simply "left open" for too long without calling commit or rollback, such as by a user taking a while to figure out how to write the SQL they need to create the infra they want, this is not something that a cron should *ever* do as it's essentially "just" `SELECT iasql_begin(); SELECT iasql_commit()`, so *perhaps* what we need is a `commit_started` audit log record we execute the timeout on, not against the `transaction_started` audit log record?

However if we make the convention be one of "when you're doing a transaction to generate a review, you should roll back your changes when done" then we side-step the 30 minute timeout issue for this PR for all but the "taking too long to write the IaSQL to be reviewed" issue, which may still be enough to need to consider it in this PR, but we could reduce scope by ignoring that for the time being?

Reply:
First, for context, it is not a "timeout" per se, what we do is that to *open* a new transaction we check if in the last 30 minutes, there's another *open* transaction or there's a `commit` running (checking a `START_COMMIT` event). If we find one we error saying there's another transaction open, if we don't we insert the new *open* transaction log. The code can be found here: https://github.com/iasql/iasql/blob/main/src/modules/iasql_functions/iasql.ts#L1854L1887

Now, I've been thinking about this. Help me walk through it. What are the possible scenarios if we remove the "timeout" and the engine crashes and/or restarts? I see 2 main scenarios:

1. Container without persistence
In this scenario, it does not matter that much. when the container restarts the data will be gone, with no audit logs and will work as a fresh container, right?

2. Container with persistent data
From this scenario I see 2 possible issues depending on which was the *latest relevant* log:

- `OPEN_TRANSACTION`: In this case, it will be good, there will be an open transaction, and the users can execute `commit` or `rollback` since these 2 methods need an open transaction to work.
- `START_COMMIT`: This case will mean some change might have gone through, but some not. Hard to know. If we try a `commit` or a `rollback` it will try to look if there's an open transaction, which will be there. The "problem" will be that previous changes, applied or not will be ignored since when a `commit` tries to go through again, it will look for changes since the previous `START_COMMIT` which will be the one that got halfway and it may not be new changes, so it will just `sync` and unlock.

Am I missing a case related to the restart/crash I'm not taking into account it could leave things in a locked state? If you have in mind a possible scenario let me know and we can walk through it.

Reply:
First, for context, it is not a "timeout" per se, what we do is that to *open* a new transaction we check if in the last 30 minutes, there's another *open* transaction or there's a `commit` running (checking a `START_COMMIT` event). If we find one we error saying there's another transaction open, if we don't we insert the new *open* transaction log. The code can be found here: https://github.com/iasql/iasql/blob/main/src/modules/iasql_functions/iasql.ts#L1854L1887

Now, I've been thinking about this. Help me walk through it. What are the possible scenarios if we remove the "timeout" and the engine crashes and/or restarts? I see 2 main scenarios:

1. Container without persistence
In this scenario, it does not matter that much. when the container restarts the data will be gone, with no audit logs and will work as a fresh container, right?

2. Container with persistent data
From this scenario I see 2 possible issues depending on which was the *latest relevant* log:

- `OPEN_TRANSACTION`: In this case, it will be good, there will be an open transaction, and the users can execute `commit` or `rollback` since these 2 methods need an open transaction to work.
- `START_COMMIT`: This case will mean some change might have gone through, but some not. Hard to know. If we try a `commit` or a `rollback` it will try to look if there's an open transaction, which will be there. Currently, we are checking in the `commit` function if there's another commit running. In this case that will fail since that commit will never end. If we add a "timeout" or bypass that one, then another "problem" will be that previous changes, applied or not will be ignored since when a `commit` tries to go through again, it will look for changes since the previous `START_COMMIT` which will be the one that got halfway and it may not be new changes, so it will just `sync` and unlock.

Am I missing a case related to the restart/crash I'm not taking into account it could leave things in a locked state? If you have in mind a possible scenario let me know and we can walk through it.

Reply:
So this issue can affect even containers without persistence in the following way:

1. `iasql_commit` is called that has triggered a lot of work.
2. That RPC call is an HTTP req under the hood to the engine, so if it was triggered by the user it apparently has a 15 minute limit from the browser, and a [1 hour limit](https://github.com/iasql/iasql/blob/main/src/modules/interfaces.ts#L554) between Postgres and the engine.
3. If the commit takes longer than an hour *and then fails* in a semi-safe way (not doing what we expected, but not crashing the process), returning an error message back up, there is nothing monitoring that error message anymore. 
4. Now the `START_COMMIT` has no `END_COMMIT` and never will, so if we just wait forever on this point, no cron job will ever run again, and the current behavior of `iasql_commit` and `iasql_rollback` may or may not function in this case, so it may not even be possible to get out of this situation.

Because the cron job uses the same audit log event types as the user, we can't be sure if we should automatically try again (if it was from the cron) or if we should wait for user input (if it was from the user). The *reason* why we shouldn't always rely on input from the user to get things un-stuck is if the user does not use transactions at all and is using it solely as an eventually-consistent database. The user may not even know the transaction option is possible or how it works under the hood, so expecting the user to get things un-stuck seems unreasonable to me.

Now, on your other point with `OPEN_TRANSACTION` but no `START_COMMIT`, this piece is *clearly* only possible if it was initiated by the user, so blocking forever in *this* state seems perfectly fine to me.

The conundrum that I see, though, is if it gets stuck for too long in `START_COMMIT` state; if it was via `cron` the proper solution is to just try again and not try to roll back the user's changes, but if it was a user-initiated transaction, catching this and automatically initiating a rollback seems more reasonable.

If we can't figure out how to distinguish between these two states, or if we don't want to (which could also be reasonable), then "compromising" by always triggering a rollback in this sort of stuck state seems less bad; the eventually consistent user would see that their change never went through, try again, and hopefully it then works, while the user that uses transactions would get the error back that the commit failed and that it rolled back for that reason and then they can decide if they want to try again or not.

But do we have these two different states right now? I thought we only had the beginning and end of the transaction, and no event on when a commit or rollback has started.

Reply:
Ideally we can use the same solution in all situations and I was imagining a full revert, but simply getting out of the broken state with a sync could also be fine -- perhaps if there's a `START_COMMIT` without an `END_COMMIT` after some timeout that triggers the cron job to execute a `sync` instead of a `commit`?

Reply:
yes, I can see that, like if the `START_COMMIT` has been open for I don't know, 5hrs, call `sync` instead? and then we can remove the logic for the open transaction. The conditional would be if it is open and how long have been "commiting".